index.bs

```diff
@@ -391,6 +391,54 @@ Implementers of this API are expected to be familiar with the <a href="https://g
 
 
 # Programming Model # {#programming-model}
+## Overview ## {#programming-model-overview}
+
+At the heart of neural networks is a computational graph of mathematical operations.
+These operations are the building blocks of modern machine learning technologies in
+computer vision, natural language processing, and robotics.
+The WebNN API is a specification for constructing, compiling, and executing computational
+graphs of neural networks.
+
+The {{MLGraph}} interface represents a compiled computational graph (that is, a model) and exposes
+a compute method to perform inference.
+
+The {{MLGraphBuilder}} interface serves as a builder (factory) to create a {{MLGraph}}.
+(In programming language terminology, a {{MLGraph}} represents a *program/function*.)
+The {{MLOperand}} interface is a static (compile-time) representation of runtime values,
+which include input-values for inference, constants (including trained weights)
+used for inference, intermediate values (often referred to as activations) computed
+during inference, as well as the output values of inference.
+(In programming language terminology, a {{MLOperand}} represents a *variable* and is
+distinct from a runtime *value*, which is represented by a tensor. During inference,
+every {{MLOperand}} will be bound to a value.)
+The {{MLGraphBuilder}} interface enables the creation of {{MLOperand}}s.
+A key part of the {{MLGraphBuilder}} interface are the operations (such as 
+{{MLGraphBuilder/gemm()}} and {{MLGraphBuilder/softmax()}}). The operations have a functional
+semantics, with no side effects.
+Each operation invocation conceptually returns a distinct new value, without
+changing the value of any other {{MLOperand}}.
+
+The {{MLGraphBuilder/build()}} method of the {{MLGraphBuilder}} interface is used to compile and optimize
+the computation graph used to compute one or more specified outputs. The key
+purpose of the compilation step is to enable optimizations that span two or
+more operations, such as operation or loop fusion.
+
+The {{MLGraph/compute()}} method of the {{MLGraph}} interface is used to execute the
+compiled computation graph (to perform inference). The caller supplies the input
+values using {{MLNamedInputs}}, binding the input {{MLOperand}}s to their values.
```

Comment:
The caller would also provide the pre-allocated buffers to output `MLOperand` by using `MLNamedOutputs`.

Reply:
Added this, thanks. There is one thing I don't understand: I assume that the ```sequence<long> dimensions``` in ```MLInput``` represents shape information. It looks like it is optional for inputs and missing for outputs. Can you clarify the intention? Thanks!

Reply:
The `dimensions` of `MLInput` is useful for the dynamic input shape use case. The call can specify the concrete input shape when compute. For output, it expects the caller to provide the pre-allocated output buffer, it assumes the caller (usually the framework) knows the output shape and allocates the output buffer.

Do you see anything missing?

Reply:
In all DNN frameworks, the Tensor abstraction combines a pointer to some memory (the data or the buffer), with a minimum of shape (the dimensions). Aside from memory-allocation, most tensor-processing is dependent on the shape information. So, it seems to me that such an abstraction would be useful in this specification too.

Reply:
> Aside from memory-allocation, most tensor-processing is dependent on the shape information.

I agree. The shape information is provided in the graph build time. At compute time, only buffer needs to be bound except for the input with dynamic shape, it needs both data and concrete shape.